lib/evmone_precompiles/bls.cpp

```diff
@@ -36,24 +36,72 @@ constexpr auto FP_BYTES_OFFSET = 64 - 48;
     return p0_affine;
 }
 
-/// Stores p1 point in two 64-bytes arrays with big endian encoding zero padded.
-void store_result(uint8_t _rx[64], uint8_t _ry[64], const blst_p1* out) noexcept
+/// Validates p2 affine point. Checks that point coordinates are from the BLS12-381 field and
+/// that the point is on curve. https://eips.ethereum.org/EIPS/eip-2537#abi-for-g2-addition
+[[nodiscard]] std::optional<blst_p2_affine> validate_p2(
+    const uint8_t _x[128], const uint8_t _y[128]) noexcept
 {
-    blst_p1_affine result;
-    blst_p1_to_affine(&result, out);
+    constexpr auto is_field_element = [](const uint8_t _p[128]) {
+        return intx::be::unsafe::load<intx::uint512>(_p) < BLS_FIELD_MODULUS &&
+               intx::be::unsafe::load<intx::uint512>(&_p[64]) < BLS_FIELD_MODULUS;
+    };
+
+    if (!is_field_element(_x))
+        return std::nullopt;
+    if (!is_field_element(_y))
+        return std::nullopt;
 
+    blst_fp x0;
+    blst_fp x1;
+    blst_fp y0;
+    blst_fp y1;
+    blst_fp_from_bendian(&x0, &_x[FP_BYTES_OFFSET]);
+    blst_fp_from_bendian(&x1, &_x[FP_BYTES_OFFSET + 64]);
+    blst_fp_from_bendian(&y0, &_y[FP_BYTES_OFFSET]);
+    blst_fp_from_bendian(&y1, &_y[FP_BYTES_OFFSET + 64]);
+
+    const blst_p2_affine p_affine{{x0, x1}, {y0, y1}};
+    if (!blst_p2_affine_on_curve(&p_affine))
+        return std::nullopt;
+
+    return p_affine;
+}
+
+/// Stores fp in 64-bytes array with big endian encoding zero padded.
+void store(uint8_t _rx[64], const blst_fp* _x) noexcept
+{
     std::memset(_rx, 0, FP_BYTES_OFFSET);
-    blst_bendian_from_fp(&_rx[FP_BYTES_OFFSET], &result.x);
-    std::memset(_ry, 0, FP_BYTES_OFFSET);
-    blst_bendian_from_fp(&_ry[FP_BYTES_OFFSET], &result.y);
+    blst_bendian_from_fp(&_rx[FP_BYTES_OFFSET], _x);
+}
+
+/// Stores fp2 in 128-bytes array with big endian encoding zero padded.
+void store(uint8_t _rx[128], const blst_fp2* _x) noexcept
+{
+    store(_rx, &_x->fp[0]);
+    store(&_rx[64], &_x->fp[1]);
 }
+
+/// Stores p1 affine point in two 64-bytes arrays with big endian encoding zero padded.
+void store(uint8_t _rx[64], uint8_t _ry[64], const blst_p1_affine* in) noexcept
```

Comment:
These helpers for storing p1/p2 points feels like too much because of the independent `_rx` and `_ry` parameters here and in the public API.